Sources/App/Models/Version.swift

```diff
@@ -54,12 +35,14 @@ final class Version: Model, Content {
     @Field(key: "commit")
     var commit: String?
 
+    // TODO: sas-2020-05-03: currently concatenating os + version - we could save the structure instead
     @Field(key: "supported_platforms")
     var supportedPlatforms: [Platform]
```

Comment:
This one is important to discuss. Currently we'll have this filled in as

```
["macos_10.15", "ios_13"]
```

(`Platform` is just a `String` typealias). I could imagine we actually want the structured data from the package dump to poke through here:

```
struct Manifest: Codable {
    struct Platform: Codable {
        enum Name: String, Codable {
            case macos
            case ios
            case tvos
            case watchos
        }
        var platformName: Name
        var version: String
    }
    var name: String
    var swiftLanguageVersions: [String]?
    var platforms: [Platform]?
}
```

and save `Manifest.Platform` here instead (i.e. as a `jsonb[]` in the database).

It'll depend a bit on how we need to query/display this info. I mean, we can still get at the bits with PG JSON queries but if we only ever intend to display those labels as fixed strings we might as well save them that way and keep the schema simpler.

Reply:
So before looking at the specifics of how this gets represented in code, I think it's worth taking a second to think about how this data is going to be used.

I think the biggest use for this data is not really about the specific version numbers, but more as an indication of general plaform support. If I pick a package, and it supports iOS, and it's been updated in the last 12 months, it almost certainly supports the version of iOS I'm interested in. I think having this data act as a boolean flag for platform support is 90% of what's needed.

I do think we should store the version requirement, but it's only for that last 10%.

As you'll have seen, these are stored as strings for each platform in the current schema:

```
supported_macos_version
supported_ios_version
supported_watchos_version
supported_tvos_version
```

I don't have a huge preference whether these are just strings, or integers mapped to enum values, but I think having them as four columns is a big advantage when it comes to search. 

To answer your original question. I can't see anything right now that would need this data in a format other than a string. So my gut feeling says just go with that? What are your thoughts?

Reply:
I think this only makes a difference if we're looking to query or aggregate across versions - which seems unlikely, at least on a frequent enough basis that it could turn into a performance issue. Even then, I've dealt with PG databases and millions of rows with JSON queries and not had issues. Plus there's indexes on JSON field properties.

All that is to say, since I expect us to pretty much mostly pull in the whole record and then show the contents, it doesn't really make a big difference whether the fields are in columns or in a JSON field.

An advantage the latter brings is that the AR glasses that'll be introduced next year won't require a schema change :D Or perhaps some way of adding linux for that matter.

What I'd envisioned is not an integer mapped enum but a JSON array that encodes the structure pretty much exactly as laid out:

```
[
	{
	  "name": "macos",
	  "version": "10.10"
	},
	{
	  "name": "ios",
	  "version": "13.0"
	}
]
```

Since we're using Vapor on both backend and frontend, in the end that's a completely transparent decode that just gives you a `[Platform]` array to work with and we don't even have to worry about the details.

You'd still just do something like

```swift
extension Array where Element == Platform {
  var supportedPlatforms: Set<Platform.Name> { Set(map(\.name)) }
}

package.versions.map {
  // some table row template pseudo code
  "<tr>\($0.tagName) \($0.supportedPlatforms)</tr>")
}
```

i.e. we put the "smarts" into `Platform` extensions client side and make the field just a decodable store. If we ever need deeper inspection of the field content relationally, PG still has us covered with its JSON support.

And if we want to add Linux, we'd just add an enum `case linux` and that's it. As long as we don't take out any cases we don't need to do anything else.

Does that make sense?

Reply:
Sorry I missed that I hadn't replied to this. I think this is fine.

> All that is to say, since I expect us to pretty much mostly pull in the whole record and then show the contents, it doesn't really make a big difference whether the fields are in columns or in a JSON field.

You're absolutely right, with one potential exception that may affect speed. Conditional search. I'm writing up some other thoughts on Search today (will post shortly) which could potentially affect this decision. Let's revisit it after having that discussion.